### PR TITLE
Use named parameters in snippets where ambiguous

### LINKED ISSUE #1841: C# possible snippet generation ambiguity on List methods
Occurs when there are two flattenings, where one has one extra string parameter.

### PR DESCRIPTION
**Pull request by chrisdunelm**
Fixes #1841

### CODE REVIEW

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java [hunk 1]
```diff
@@ -134,14 +134,49 @@ private SnippetsFileView generateSnippets(GapicInterfaceContext context) {
       } else if (methodConfig.isPageStreaming()) {
         if (methodConfig.isFlattening()) {
           ImmutableList<FlatteningConfig> flatteningGroups = methodConfig.getFlatteningConfigs();
+          // Find flattenings that have ambiguous parameters, and mark them to use named parameters.
```
Comment (jskeet) [inline src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java]: Nit: named arguments, not named parameters. (Including variable name, method names etc.)
I know it's a bit of a nit, but I'd rather we got it right from the start...

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java [hunk 2]
```diff
@@ -134,14 +134,49 @@ private SnippetsFileView generateSnippets(GapicInterfaceContext context) {
...
+          // These are page-stream flattenings that have one or two string parameters (that are not
+          // resource-names) than another flattening.
+          String[] stringParams =
+              flatteningGroups
+                  .stream()
+                  .map(
+                      flat ->
```
Comment (jskeet) [inline src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java]: Is this "lambda parameter on its own" format a common one in Java? It looks very sparse to me...

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java [hunk 3]
```diff
@@ -134,14 +134,49 @@ private SnippetsFileView generateSnippets(GapicInterfaceContext context) {
       } else if (methodConfig.isPageStreaming()) {
         if (methodConfig.isFlattening()) {
           ImmutableList<FlatteningConfig> flatteningGroups = methodConfig.getFlatteningConfigs();
+          // Find flattenings that have ambiguous parameters, and mark them to use named parameters.
+          // These are page-stream flattenings that have one or two string parameters (that are not
+          // resource-names) than another flattening.
```
Comment (jskeet) [inline src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java]: Not sure what "than another" means here.

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java [hunk 4]
```diff
@@ -134,14 +134,49 @@ private SnippetsFileView generateSnippets(GapicInterfaceContext context) {
...
+                                      field.getField().getType().isStringType()
+                                              && field.getResourceNameConfig() == null
+                                          ? 's'
+                                          : '.')
+                              .collect(
+                                  StringBuilder::new,
+                                  StringBuilder::appendCodePoint,
+                                  StringBuilder::append)
+                              .toString())
+                  .toArray(String[]::new);
```
Comment (jskeet) [inline src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java]: I have very little idea what the result is here. Is this code that exists somewhere else for another reasons?

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java (hunk 1 above)
Comment (chrisdunelm) [reply src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java]: All changed.

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java (hunk 4 above)
Comment (chrisdunelm) [reply src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java]: Added more comments to try to explain. No, this code was written here specifically for this.

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java (hunk 3 above)
Comment (chrisdunelm) [reply src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java]: I've tried to explain better.

File: src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java (hunk 2 above)
Comment (chrisdunelm) [reply src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java]: (Thought I'd already answered this...)

This code is auto-formatted, and the formatting is enforced, so the sparseness cannot be changed,

**Pull request merged**